scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1587 @@
+
+    Parameters
+    ----------
+    d : int
+        Dimension of the parameter space.
+    seed : {int or `numpy.random.RandomState` instance}, optional
+        If `seed` is not specified, the `numpy.random.RandomState`
+        singleton is used.
+        If `seed` is an int, a new ``RandomState`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``RandomState`` instance, then that
+        instance is used.
+
+    Notes
+    -----
+    By convention samples are distributed over the half-open interval
+    ``[0, 1)``. Instances of the class can access the attributes: ``d`` for
+    the dimension; and ``rng`` for the random number generator (used for the
+    ``seed``).
+
+    **Subclassing**
+
+    When subclassing `QMCEngine` to create a new sampler,  ``__init__`` and
+    ``random`` must be redefined.
+
+    * ``__init__(d, seed=None)``: at least fix the dimension. If the sampler
+      does not take advantage of a ``seed`` (deterministic methods like
+      Halton), this parameter can be omitted.
+    * ``random(n)``: draw ``n`` from the engine.
+
+    Optionally, two other methods can be overwritten by subclasses:
+
+    * ``reset``: Reset the engine to it's original state.
+    * ``fast_forward``: If the sequence is deterministic (like Halton
+      sequence), then ``fast_forward(n)`` is skipping the ``n`` first draw.
+
+    Examples
+    --------
+    To create a random sampler based on ``np.random.random``, we would do the
+    following:
+
+    >>> from scipy.stats import qmc
+    >>> class RandomEngine(qmc.QMCEngine):
+    ...     def __init__(self, d, seed):
+    ...         super().__init__(d=d, seed=seed)
+    ...         self.rng_seed_state = self.rng.get_state()
+    ...
+    ...
+    ...     def random(self, n=1):
+    ...         return self.rng.random((n, self.d))
+    ...
+    ...
+    ...     def reset(self):
+    ...         self.rng.set_state(self.rng_seed_state)
+    ...         return self
+    ...
+    ...
+    ...     def fast_forward(self, n):
+    ...         self.rng.random((n, self.d))
+    ...         return self
+
+    After subclassing `QMCEngine` to define the sampling strategy we want to use,
+    we can create an instance to sample from.
+
+    >>> engine = RandomEngine(2, seed=12345)
+    >>> engine.random(5)
+    array([[0.92961609, 0.31637555],
+           [0.18391881, 0.20456028],
+           [0.56772503, 0.5955447 ],
+           [0.96451452, 0.6531771 ],
+           [0.74890664, 0.65356987]])
+
+    We can also reset the state of the generator and resample again.
+
+    >>> _ = engine.reset()
+    >>> engine.random(5)
+    array([[0.92961609, 0.31637555],
+           [0.18391881, 0.20456028],
+           [0.56772503, 0.5955447 ],
+           [0.96451452, 0.6531771 ],
+           [0.74890664, 0.65356987]])
+
+    """
+
+    @abstractmethod
+    def __init__(self, d: int,
+                 seed: Optional[int, np.random.RandomState] = None):
```

Comment:
`Generator` here too.

Also, `__init__` methods should be annotated with `-> None`

Reply:
This applies further down in other classes as well.